keps/sig-cluster-lifecycle/wgs/20200108-instance-specific-componentconfig.md

```diff
@@ -0,0 +1,644 @@
+instance of the component.
+
+Other components may not be deployed as Pods but may still have channels to
+provide configuration via the cluster. Kubelet, for example, has the Dynamic
+kubelet Config feature, which enables the kubelet to read configuration from
+a ConfigMap designated by its corresponding Node object. With kubelet, it is
+also advantageous for "pools" of nodes to refer to the same ConfigMap.
+
+Unfortunately, a shared ConfigMap is incompatible with instance-specific
+parameters which, by definition, cannot be shared between instances. While we
+have been able to migrate many shareable parameters to ComponentConfig APIs,
+we have been blocked on the instance-specific piece. Lack of a decision on how
+to handle these has led to uncertainty as the owners of other components begin
+moving to ComponentConfig APIs. A clear specification on how to implement
+instance-specific configuration will unblock and expedite our migration from
+command-line flags to ComponentConfig.
+
+See also issue [#61647](https://github.com/kubernetes/kubernetes/issues/61647).
+
+### Goals
+
+- Provide a clear specification on how to handle instance-specific in
+  ComponentConfig when migrating to a new or adapting an existing
+  ComponentConfig API.
+
+### Non-Goals
+
+- Provide a standard way to continue specifying configuration on the
+  command-line.
+- Redesign APIs/refactor the existing set of parameters. This is about providing
+  a clear migration path for all classes of parameters.
+- Eliminate instance-specific config parameters altogether. This is a special
+  case of the previous item.
+- Make changes to the Dynamic kubelet Config feature or address difficulties in
+  using the kubelet's current ComponentConfig API with Dynamic kubelet Config.
+
+## Proposal
+
+Taking the Kubelet as an example, there are three categories of fields
+(identified by @liggitt):
+1. clearly instance-specific (e.g. `--node-ip`, `--provider-id`,
+   `--hostname-override`)
+2. possibly instance-specific (e.g. `--bind-address`, `--root-dir`)
+3. highly unlikely to be instance-specific (e.g. cloud provider, node lease
+   duration)
+
+This KEP proposes that components have a way to define the same field in
+both instance-specific and shareable configuration files, defining a generic
+merge approach such that the instance-specific field overrides and takes
+precedence over the shareable field.
+
+### Reuse component config Kind for instance-specific config
+
+This KEP proposes a new flag, `--instance-config`, to specify the instance-specific
+configuration file, the instance-specific configuration has the same Kind as the
+sharable one.
+
+ComponentConfig files containing only sharable parameters can continue to be
+shared via a single ConfigMap. Files containing only instance-specific
+parameters can be provided to components via other means, such as the node
+startup scripts or an init container that inserts values from the Pod's Downward
+API.
```

Comment:
i must admit i'm not really fond of the idea of reusing the Kinds with separate flags.
this leaves Kinds to hold a sparse collection of global and instance fields, which can confuse users unless each field is documented and requires separate flags.

have you considered?
```
kind: SomeKind
instance:
 ...
global:
 ...
```

this can be passed via a singular --config.

Reply:
Is "you can override whatever you want locally" really that confusing?

There's also a high chance we'll screw up the categorization of some field and create additional friction for users, when instead we can solve this up-front (and avoid future debates) by giving users flexibility. If users don't need the flexibility, they don't have to opt-in to --instance-config.

Reply:
I think the key idea here is that "instance specific" is contextual and depends on the user's environment, and we shouldn't force them into a corner.

Reply:
how would global fields be handled if passed via --instance-config. does that mean the component has to omit patching certain fields when merging instance over global?

if the global configuration already exists in a ConfigMap, wouldn't passing --config with instance specific parameters act the same as passing --instance-config?

Reply:
--instance-config means "an override local to this instance." The component doesn't have to do anything special to categorize fields as instance-specific or not. It should just treat --instance-config as a patch over --config.

If the global config is already present in a file, in this case delivered via a ConfigMap, and instance-specific overrides are required, this KEP recommends creating a parallel patch file that can be specified via --instance-config, and treating the file delivered via the ConfigMap as an immutable base config. 

Directly modifying a file delivered via a ConfigMap is an antipattern, because it makes that file out-of-sync with the ConfigMap.

Reply:
if component Foo has a global setting such as a cluster wide CIDR, but the --instance-config overrides it (with an unfiltered patch) for that particular instance, i'm assuming that can create problems for the behavior of that particular component instance.

Reply:
Yes, but I think we should trust users on that. I don't expect a user to reach for --instance-config until it turns out that --config isn't sufficient for whatever they need.